### ISSUE TITLE
Add support for `signTypedData` in @truffle/hdwallet-provider

### ISSUE THREAD

**Issue by dmihal** [trufflesuite/truffle#3585]
## Issue

When sending an `eth_signTypedData` call, the provider throws the following error:

```
     TypeError: self.signTypedMessage is not a function
      at waterfall (node_modules/@trufflesuite/web3-provider-engine/subproviders/hooked-wallet.js:390:18)
```

## Steps to Reproduce

I've added a simple test to the `eth-permit` repo.

```
git clone https://github.com/dmihal/eth-permit.git
git checkout hdprovider
yarn install
yarn test

```

## Fix

I believe this issue is fixed by adding signTypedMessage to the HookedSubprovider.

**Comment by gnidan** [comment]
Thanks for opening this @dmihal. To the best of our knowledge, this just needs to be added as an enhancement. Putting this on our backlog so that we can get to doing this (also labeling it help wanted in case anyone wants to open the PR themselves ;)

**Comment by lucas-goldner** [comment]
Any updates on this I really need to get this working at the moment ? Any "fix" already available ?

**Comment by danfinlay** [comment]
These methods as provided by MetaMask are available on [eth-sig-util](https://github.com/MetaMask/eth-sig-util), so that would be a useful library for implementing this support.

**Comment by gabmontes** [comment]
It looks like `signTypedMessage` must be provided to the constructor of `@trufflesuite/web3-provider-engine`'s `HookedWalletSubprovider`:

https://github.com/trufflesuite/provider-engine/blob/master/subproviders/hooked-wallet.js#L82

The subprovider is instantiated by `@truffle/hdwallet-provider` here:

https://github.com/trufflesuite/truffle/blob/%40truffle/hdwallet-provider%401.2.3/packages/hdwallet-provider/src/index.ts#L133

But it looks there is no [clean] way to add an additional `signTypedMessage` there...

**Comment by selimerunkut** [comment]
is there a quick fix for this, stuck and can't deploy the contracts

**Comment by selimerunkut** [comment]
With truffle test tests, I use this function and everything is signed fine:

```
signTypedData: function (web3, from, data) {
    return new Promise(async (resolve, reject) => {
      function cb(err, result) {
        if (err) {
          return reject(err);
        }
        if (result.error) {
          return reject(result.error);
        }

        const sig = result.result;
        const sig0 = sig.substring(2);
        const r = "0x" + sig0.substring(0, 64);
        const s = "0x" + sig0.substring(64, 128);
        const v = parseInt(sig0.substring(128, 130), 16);

        resolve({
          data,
          sig,
          v, r, s
        });
      }
      if (web3.currentProvider.isMetaMask) {
        web3.currentProvider.sendAsync({
          jsonrpc: "2.0",
          method: "eth_signTypedData_v3",
          params: [from, JSON.stringify(data)],
          id: new Date().getTime()
        }, cb);
      } else {
        let send = web3.currentProvider.sendAsync;
        if (!send) send = web3.currentProvider.send;
        send.bind(web3.currentProvider)({
          jsonrpc: "2.0",
          method: "eth_signTypedData",
          params: [from, data],
          id: new Date().getTime()
        }, cb);
      }
    }); 
```

**Comment by eggplantzzz** [comment]
Thanks @dmihal for the test project. There is currently a PR to implement support for this [here](https://github.com/trufflesuite/truffle/pull/4124) but I'm having trouble validating the fix with your project. This is because in the implementation of `signTypedMessage` in hdwallet-provider it expects the "data" param to be an object while you send the "data" param as stringified JSON. On searching around, I see examples of it being sent to the provider both ways, which one is correct? Can you provide me with any resources about this?

**Comment by gnidan** [comment]
This functionality is released in [Truffle v5.4.7](https://github.com/trufflesuite/truffle/releases/tag/v5.4.7). Thanks all!
